### PR TITLE
feat(application-templates-party-application): Subsystem authentication scopes

### PR DESCRIPTION
**Pull request by stefanhar**
# Authentication scopes for party application subsystems

Currently subsystems used by paty application lack authentication scopes.
This PR adds the required scopes and assigns required scopes to authentication for user facing systems

## What

- Added scopes to endorsement system
- Added scopes to party letter registry
- Added scopes to temporary voter registry
- Assigned scopes to application system authentication
- Assigned scopes to service portal authentication
- Updated endpoints to extract national id from auth instead parameters

## Why

- Limit access to endpoints by unwanted parties

## Checklist:

- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: apps/services/endorsements/api/src/app/modules/endorsementList/endorsementList.controller.ts [hunk 1]
```diff
@@ -22,11 +21,12 @@ import { EndorsementListService } from './endorsementList.service'
 import { EndorsementListDto } from './dto/endorsementList.dto'
 import { FindEndorsementListByTagsDto } from './dto/findEndorsementListsByTags.dto'
 import { Endorsement } from '../endorsement/endorsement.model'
-import { BypassAuth, CurrentUser } from '@island.is/auth-nest-tools'
+import { BypassAuth, CurrentUser, Scopes } from '@island.is/auth-nest-tools'
 import { EndorsementListByIdPipe } from './pipes/endorsementListById.pipe'
 import { IsEndorsementListOwnerValidationPipe } from './pipes/isEndorsementListOwnerValidation.pipe'
 import { environment } from '../../../environments'
 import type { User } from '@island.is/auth-nest-tools'
+import { Endorsement as EndorsementScope } from '@island.is/auth/scopes'
```
Comment (saevarma) [inline apps/services/endorsements/api/src/app/modules/endorsementList/endorsementList.controller.ts:29]: See comment below about renaming the scopes enums to end with Scope.

File: apps/services/endorsements/api/src/app/modules/endorsement/endorsement.controller.ts [hunk 2]
```diff
@@ -27,6 +26,8 @@ import { IsEndorsementListOwnerValidationPipe } from '../endorsementList/pipes/i
 import { BulkEndorsementDto } from './dto/bulkEndorsement.dto'
 import { Endorsement } from './endorsement.model'
 import { EndorsementService } from './endorsement.service'
+import { Endorsement as EndorsementScope } from '@island.is/auth/scopes'
```
Comment (saevarma) [inline apps/services/endorsements/api/src/app/modules/endorsement/endorsement.controller.ts:29]: See comment below about renaming the scopes enums to end with Scope.

**Review by saevarma — CHANGES_REQUESTED** [review]
Few comments regarding the scopes and usage of the Audit decorator.

File: libs/auth/scopes/src/lib/endorsement.scope.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+export enum Endorsement {
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply libs/auth/scopes/src/lib/endorsement.scope.ts:1]: Missed that, will update

File: libs/auth/scopes/src/lib/partyLetterRegistry.scope.ts [hunk 4]
```diff
@@ -0,0 +1,4 @@
+export enum PartyLetterRegistry {
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply libs/auth/scopes/src/lib/partyLetterRegistry.scope.ts:1]: Missed that, will update

File: apps/services/temporary-voter-registry-api/src/environments/environment.ts [hunk 5]
```diff
@@ -1,3 +1,31 @@
...
+  },
+  audit: {
+    defaultNamespace: '@island.is/services-temporary-voter-registry-api',
+  },
 }
+
+const prodConfig = {
+  auth: {
+    issuer: process.env.IDENTITY_SERVER_ISSUER_URL,
+    audience: '',
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/temporary-voter-registry-api/src/environments/environment.ts:18]: Roger, will add that audience

File: apps/services/temporary-voter-registry-api/src/openApi.ts [hunk 6]
```diff
@@ -1,8 +1,22 @@
...
   .setVersion('1.0')
   .addTag('temporaryVoterRegistry')
+  .addOAuth2({
+    type: 'oauth2',
+    flows: {
+      authorizationCode: {
+        authorizationUrl: environment.swagger.authUrl,
+        tokenUrl: environment.swagger.tokenUrl,
+        scopes: {
+          'openid profile':
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/temporary-voter-registry-api/src/openApi.ts:16]: Ty, nice catch 👍

File: apps/services/temporary-voter-registry-api/src/app/modules/voterRegistry/voterRegistry.controller.ts [hunk 7]
```diff
@@ -1,34 +1,47 @@
...
   @ApiOkResponse({
-    description: 'Finds voters region given voters national id',
+    description: 'Finds voter registry entry given user authentication',
     type: VoterRegistry,
   })
+  @Audit<VoterRegistry>({
+    namespace: auditNamespace,
+    action: 'findByAuth',
+    resources: (voterRegistry) => voterRegistry.id,
+  })
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/temporary-voter-registry-api/src/app/modules/voterRegistry/voterRegistry.controller.ts:26]: Love it

File: apps/services/temporary-voter-registry-api/src/app/modules/voterRegistry/voterRegistry.controller.ts [hunk 8]
```diff
@@ -1,34 +1,47 @@
...
 
-    return resource
+  // TODO: This should get a system scope, or we should allow systems to get the read scope
+  // TODO: Add tests for this once its design is finalized
+  @ApiOkResponse({
+    description: 'Finds voter registry entry given user authentication',
+    type: VoterRegistry,
+  })
+  @Get('/system')
+  async findByNationalId(
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/temporary-voter-registry-api/src/app/modules/voterRegistry/voterRegistry.controller.ts:42]: It's unclear to us how this endpoint should be handled, currently we have no audit/scopes defined

File: apps/services/temporary-voter-registry-api/src/app/modules/voterRegistry/e2e/findByAuth/findByAuth.spec.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        .overrideProvider(IdsUserGuard)
+        .useValue(
+          new MockAuthGuard({
+            nationalId,
+            scope,
+          }),
+        )
+        .compile()
+    },
+  })
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/temporary-voter-registry-api/src/app/modules/voterRegistry/e2e/findByAuth/findByAuth.spec.ts:31]: Clearly, must have forgotten to move it 😅

File: apps/services/party-letter-registry-api/src/environments/environment.ts [hunk 10]
```diff
@@ -1 +1,31 @@
-export default {}
+const devConfig = {
+  auth: {
+    issuer: 'https://identity-server.dev01.devland.is',
+    audience: '',
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/party-letter-registry-api/src/environments/environment.ts:4]: Roger, will add that audience

File: apps/services/party-letter-registry-api/src/app/modules/partyLetterRegistry/e2e/findByManager/findByManager.spec.ts [hunk 11]
```diff
@@ -1,43 +1,47 @@
...
 import { errorExpectedStructure } from '../../../../../../test/testHelpers'
 import * as request from 'supertest'
-import { INestApplication } from '@nestjs/common'
-
-let app: INestApplication
-
-beforeAll(async () => {
-  app = await setup()
-})
+import { PartyLetterRegistry as PartyLetterRegistryScope } from '@island.is/auth/scopes'
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/party-letter-registry-api/src/app/modules/partyLetterRegistry/e2e/findByManager/findByManager.spec.ts:4]: This should have been a giveaway tbh 😅

File: apps/services/endorsements/api/src/app/modules/endorsementList/endorsementList.controller.ts (hunk 1 above)
Comment (stefanhar) [reply apps/services/endorsements/api/src/app/modules/endorsementList/endorsementList.controller.ts:29]: Roge, will update

File: apps/services/endorsements/api/src/app/modules/endorsement/endorsement.controller.ts (hunk 2 above)
Comment (stefanhar) [reply apps/services/endorsements/api/src/app/modules/endorsement/endorsement.controller.ts:29]: Will update 👍

File: libs/auth/scopes/src/lib/endorsement.scope.ts [hunk 12]
```diff
@@ -0,0 +1,7 @@
+export enum EndorsementScope {
+  endorsementRead = '@island.is/endorsement/endorsement:read',
+  endorsementWrite = '@island.is/endorsement/endorsement:write',
+  endorsementListRead = '@island.is/endorsement/endorsement-list:read',
+  endorsementListWrite = '@island.is/endorsement/endorsement-list:write',
+  endorsementListJusticeDepartmentWrite = '@island.is/endorsement/endorsement-list-justice-department:write',
+}
```
Comment (saevarma) [inline libs/auth/scopes/src/lib/endorsement.scope.ts:7]: I would consider renaming the scopes to make it a bit cleaner.
```suggestion
export enum EndorsementScope {
  read = '@island.is/endorsement:read',
  write = '@island.is/endorsement:write',
  listRead = '@island.is/endorsement/list:read',
  listWrite = '@island.is/endorsement/list:write',
  listJusticeDepartmentWrite = '@island.is/endorsement/list/justice-department:write',
}
```

File: apps/services/temporary-voter-registry-api/src/openApi.ts [hunk 13]
```diff
@@ -1,8 +1,22 @@
...
   .setTitle('Temporary voter registry')
   .setDescription('This api contains access to the temporary voter registry.')
   .setVersion('1.0')
   .addTag('temporaryVoterRegistry')
+  .addOAuth2({
+    type: 'oauth2',
+    flows: {
+      authorizationCode: {
+        authorizationUrl: environment.swagger.authUrl,
+        tokenUrl: environment.swagger.tokenUrl,
```
Comment (saevarma) [inline apps/services/temporary-voter-registry-api/src/openApi.ts:15]: I would consider to link this to the actual IdP by using `environment.auth.issuer`.
That way. the OpenApi will be correct in each environment and easy to change the IdP host in the future.
```suggestion
        authorizationUrl: `${environment.auth.issuer}/connect/authorize`,
        tokenUrl: `${environment.auth.issuer}/connect/token`,
```

File: apps/services/party-letter-registry-api/src/openApi.ts [hunk 14]
```diff
@@ -1,8 +1,22 @@
...
   .setTitle('Party letter registry')
   .setDescription('This api manages access to the party letter registry.')
   .setVersion('1.0')
   .addTag('partyLetterRegistry')
+  .addOAuth2({
+    type: 'oauth2',
+    flows: {
+      authorizationCode: {
+        authorizationUrl: environment.swagger.authUrl,
+        tokenUrl: environment.swagger.tokenUrl,
```
Comment (saevarma) [inline apps/services/party-letter-registry-api/src/openApi.ts:15]: Ditto

**Review by saevarma — APPROVED** [review]
LGTM. 
Only few considerations for improvements.

File: apps/services/endorsements/api/src/app/app.module.ts [hunk 15]
```diff
@@ -27,6 +28,10 @@ import { SequelizeConfigService } from './sequelizeConfig.service'
       useExisting: IdsUserGuard,
     },
     IdsUserGuard, // allows test module to see this provider for mocking auth
+    {
+      provide: 'APP_GUARD',
```
[the remark this answers is not in the document]
Comment (stefanhar) [reply apps/services/endorsements/api/src/app/app.module.ts]: Yes, solid 👍

File: libs/auth/scopes/src/lib/endorsement.scope.ts (hunk 12 above)
Comment (stefanhar) [reply libs/auth/scopes/src/lib/endorsement.scope.ts]: Agree much cleaner, will update

File: apps/services/temporary-voter-registry-api/src/openApi.ts (hunk 13 above)
Comment (stefanhar) [reply apps/services/temporary-voter-registry-api/src/openApi.ts]: Good idea, will change

**Review by siggakogk — APPROVED** [review]
lgtm

**Review by thordurhhh — APPROVED** [review]
service-portal changes LGTM 👍

File: apps/application-system/form/src/auth.ts [hunk 16]
```diff
@@ -19,7 +22,7 @@ if (userMocked) {
     redirectPathSilent: '/silent/signin-oidc',
     authority: environment.identityServer.authority,
     client_id: 'island-is-1',
-    scope: `openid profile api_resource.scope ${ApplicationScope.read} ${ApplicationScope.write} ${UserProfileScope.read} ${NationalRegistryScope.individuals}`,
+    scope: `openid profile api_resource.scope ${ApplicationScope.read} ${ApplicationScope.write} ${UserProfileScope.read} ${NationalRegistryScope.individuals} ${TemporaryVoterRegistryScope.read} ${PartyLetterRegistryScope.read} ${EndorsementScope.listWrite} ${EndorsementScope.listRead} ${EndorsementScope.read} ${EndorsementScope.write}`,
```
Comment (jeremybarbet) [inline apps/application-system/form/src/auth.ts:25]: I'm wondering if converting this for an array would be more readable and less repetitive of string interpolation and using a join to make the string within the auth library

**Review by jeremybarbet — APPROVED** [review]
LGTM

**Pull request merged**